### PR TITLE
Git spec handling revamp for 3.7

### LINKED ISSUE #14831: git url parsing unit test failure with golang-1.8
When using golang 1.8, there are failures related to parsing git urls due to changes in url.Parse in golang 1.8

##### Version
master

##### Steps To Reproduce
1. OS_BUILD_ENV_GOLANG=1.8 hack/env make release
2. OS_BUILD_ENV_GOLANG=1.8 hack/env make test-unit

##### Current Result
```
--- FAIL: TestParseRepository (0.08s)
	git_test.go:101: ParseRepository returned err: parse git@[2001:db8::1]:repository.git: first path segment in URL cannot contain colon
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x229c48]

goroutine 7 [running]:
testing.tRunner.func1(0xc42007d110)
	/usr/lib/golang/src/testing/testing.go:622 +0x25c
panic(0x274e40, 0x429700)
	/usr/lib/golang/src/runtime/panic.go:489 +0x284
github.com/openshift/origin/pkg/generate/git.TestParseRepository(0xc42007d110)
	/root/origin/_output/local/go/src/github.com/openshift/origin/pkg/generate/git/git_test.go:108 +0xf58
testing.tRunner(0xc42007d110, 0x2d7278)
	/usr/lib/golang/src/testing/testing.go:657 +0x9c
created by testing.(*T).Run
	/usr/lib/golang/src/testing/testing.go:697 +0x270
FAIL	github.com/openshift/origin/pkg/generate/git	0.137s
```

##### Expected Result
Unit test cases in pkg/generate/git/git_test.go should pass.

### LINKED ISSUE #15005: Migration of prod cluster blocked on builds and build configs that fail Git URI validation
In a prod data set being tested on 3.6

```
error:     -n XXXX buildconfigs/XXXXX: BuildConfig "XXXXX" is invalid: spec.source.git.uri: Invalid value: "git@github.com:REPO/NAME.git": uri is not a valid url
```

Our URI must support `git@SERVER:PATH`.

This blocks underlying migration of data in 3.6, and may have broken end users.

@jupierce @bparees we cannot promote 3.6 to prod without resolving this or determining it is not an issue.

### PR DESCRIPTION
**Pull request by jim-minter**
improve git repository spec parsing and hopefully avoid hitting issues due to tightened URL parsing in go 1.8.

fix #14831 
fix #15005

### CODE REVIEW

**Comment by jim-minter** [comment]
@coreydaley 
@openshift/devex 
@bparees not sure whether a conversation is warranted about if this goes in 3.6.0 or not

[test][testextended][extended:core(builds)]

**Comment by jim-minter** [comment]
@coreydaley apologies, this was a resurrection of a patch I already had; didn't realise that the above issues were in your name before I sent the PR

**Comment by coreydaley** [comment]
@jim-minter Well I am certainly not going to complain about you fixing them ;)

File: pkg/generate/git/git.go [hunk 1]
```diff
@@ -5,35 +5,64 @@ import (
...
+		u := &url.URL{
+			Scheme:   "ssh",
+			Host:     m[2],
+			Path:     m[3],
+			Fragment: m[4],
+		}
+		if m[1] != "" {
+			u.User = url.User(m[1])
+		}
+		return u, nil
```
Comment (gabemontero) [inline pkg/generate/git/git.go]: Is there similar logic now in the s2i code base as well (admissions, I could search, but am technically on PTO :-) )?  Not that I had a lot of pride in the name `MungeNoProtocolURL` :-) , but part of the original motivation for the structure of this change was to have consistent URL parsing between origin and s2i, where we pushed as much logic into s2i as possible.  Having this logic written one time and shared in both repos still seems valuable, no?

Comment (gabemontero) [inline pkg/generate/git/git.go]: To clarify, given the associated removal in the s2i repo, I'm just curious as to why it was no longer getting called there as well (but it has been a while ... perhaps I'm forgetting a reason why it was not needed there).

Comment (coreydaley) [inline pkg/generate/git/git.go]: @gabemontero I've got this issue open (https://github.com/openshift/source-to-image/issues/206) to do some refactoring/sharing of code between origin and source-to-image, which maybe should be expanded to cover as much as possible?

Comment (gabemontero) [inline pkg/generate/git/git.go]: @coreydaley yeah if we are not going to address placing this fix somehow into s2i via  https://github.com/openshift/source-to-image/pull/771, then either doing so subsequently in a common fashion, or determining/documenting why it is not needed, seems prudent to me.

**Comment by bparees** [comment]
this is a step in the right direction, but the build logic itself still has other git-url managing logic (shared with s2i, as @gabemontero alluded to, though neither invoke the munge function anymore).

Specifically this path will need to be cleaned up as well since it ultimately relies on url.Parse:
https://github.com/openshift/origin/blob/master/pkg/build/builder/source.go#L170
https://github.com/openshift/source-to-image/blob/master/pkg/scm/git/git.go#L104-L131
https://github.com/openshift/source-to-image/blob/master/pkg/scm/git/git.go#L187-L209
https://github.com/openshift/source-to-image/blob/master/pkg/scm/git/git.go#L306-L373

**Comment by bparees** [comment]
hm, maybe that's incorrect since ParseRepository will be called before any of that logic, thus ensuring a clean url reaches that code path. 

however s2i itself (cli invocation) will still be broken, i think.

**Comment by bparees** [comment]
actually i was right the first time.  this fixes the ParseRepository invocation called at one point in the build path, but the fetchSource logic that actually clones the repo does not go through ParseRepository today, so it will still break there.

File: pkg/generate/git/git.go (hunk 1 above)
Comment (bparees) [inline pkg/generate/git/git.go]: it looks like Munge* was not being called at all in s2i, i think we must have refactored around it at some point and not removed it.  So the removal isn't a concern, but the fact that s2i itself still uses url.Parse in ways that i think will break, is.

Comment (gabemontero) [inline pkg/generate/git/git.go]: 100% agreed.  Having a single code path shared by both scenarios that addresses the multitude of pain points wrt URL parsing in golang should be the ultimate goal, and what I wanted (but may not have entirely succeeded) to convey.  I'll defer on whether it is done in one fell swoop or over a series of changes.

**Comment by jim-minter** [comment]
@bparees I don't follow you - I think that url.Parse is only called in the above code path after the ssh case has been ruled out, therefore the code is "OK".

I do agree that s2i and origin appear to require similar logic here and that given the practice is to vendor in to origin, this new code is probably in the wrong place.

I guess I was aiming at taking a single small step in order to enable Go 1.8.  In principle I think some code like this should replace ParseFile, ParseSSH, ParseURL, but because of their side-effects it's a more involved change.  Let me take another look.

**Comment by jim-minter** [comment]
ah, there's a bug with relative file paths too, hold on.

**Comment by bparees** [comment]
> @bparees I don't follow you - I think that url.Parse is only called in the above code path after the ssh case has been ruled out, therefore the code is "OK".

ah, ok yes that's true.  it still looks dangerous and i'd like to see us get to a more centralized flow like what you're doing here, for anything that's processing a git spec.

> I guess I was aiming at taking a single small step in order to enable Go 1.8. In principle I think some code like this should replace ParseFile, ParseSSH, ParseURL, but because of their side-effects it's a more involved change. Let me take another look.

Fair enough, but we've got (some) time since we're not going to try to jam this in 3.6.0, so i think it's worth seeing if we can get this all refactored at once (Assuming, after your investigation, there is still refactoring to do..but i think there is, because we are manipulating the raw git spec value in two different codepaths currently)

**Comment by jim-minter** [comment]
Agreed.

**Comment by liggitt** [comment]
cc @deads2k

**Comment by jim-minter** [comment]
@bparees ready for review - need to update s2i commit SHA and squash before release, otherwise it's passing tests

**Comment by bparees** [comment]
@jim-minter lgtm, so overall the dance that needs to happen here:

1) get the s2i changes into both master and 3.6 branches of the s2i repo
2) update this PR w/ the s2i master bump commit, this PR will merge after the 3.6 branch is cut (so it's gonna be a bit, unfortunately, but discussions have started about branching).
3) create another PR against the 3.6 branch (once it exists) that includes these changes and a bump commit from the s2i 3.6 branch, so that we can get this into 3.6.1 (but *not* 3.6.0).

@deads2k will then want to pull in 2 or 3 for his rebase effort, depending on the target of that rebase.

**Comment by jim-minter** [comment]
Thanks @bparees .  Steps 1 and 2 done (2 assuming that https://github.com/openshift/source-to-image/pull/771 is the next thing to be committed to s2i master).

**Comment by jim-minter** [comment]
@deads2k I believe you should be able to cherry pick these for the rebase.

**Comment by deads2k** [comment]
Wow, no conflicts with the rebase.  It was meant to be.

**Comment by liggitt** [comment]
> no conflicts with the rebase. It was meant to be

It's a Christmas miracle in July

**Comment by jim-minter** [comment]
> `The BuildConfig "sample-build" is invalid: `
> `* spec.source.git.httpproxy: Invalid value: "127.0.0.1:3128": proxy is not a valid url`
> `* spec.source.git.httpsproxy: Invalid value: "127.0.0.1:3128": proxy is not a valid url`
>
> @jim-minter looks like we have another case of using url parse that has been broken by golang 1.8. Think you can throw together a commit for @deads2k to cherry-pick here?
>
> (https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_extended/863/)

@bparees please can you take a look at the new additional commit e50bbe9 on this PR to resolve the above?

**Comment by bparees** [comment]
> I'll be the annoying one and ask: do we need this code in s2i as well in case proxy url's are supplied there?

The standalone s2i does not handle proxies or accept a proxy url.  It assumes you've setup your git environment properly so that cloning is possible directly.

**Comment by bparees** [comment]
lgtm pending passing tests.  @deads2k once tests pass you'll want to grab the new commit into your rebase PR and then the build extended tests should pass for you.

**Comment by gabemontero** [comment]
On Tue, Jul 18, 2017 at 10:21 AM, Ben Parees <notifications@github.com>
wrote:

> I'll be the annoying one and ask: do we need this code in s2i as well in
> case proxy url's are supplied there?
>
> The standalone s2i does not handle proxies or accept a proxy url. It
> assumes you've setup your git environment properly so that cloning is
> possible directly.
>
ok cool thx for the clarification.

**Comment by deads2k** [comment]
> lgtm pending passing tests. @deads2k once tests pass you'll want to grab the new commit into your rebase PR and then the build extended tests should pass for you.

Is there a specific commit to pick?  If I have to rebase the entire thing, probably post-rebase.

**Comment by jim-minter** [comment]
@deads2k assuming it passes tests, daea5d6 on this PR is the additional commit.

**Comment by deads2k** [comment]
This didn't test green, so I did not pick it into https://github.com/openshift/origin/pull/15234

**Comment by bparees** [comment]
looks like it needs bindata generated.

**Comment by jim-minter** [comment]
bindata updated.  Most of this PR is now in via being cherry-picked; what's left is a missed Godeps update from the cherry pick and the fixes to proxy url parsing

**Comment by bparees** [comment]
[merge][severity:blocker]

**Comment by bparees** [comment]
[merge]

**Comment by bparees** [comment]
not sure why this is claiming a merge conflict, Github doesn't think it is...but @jim-minter please rebase when you get a chance.

```
      "Auto-merging pkg/build/util/util.go", 
        "CONFLICT (content): Merge conflict in pkg/build/util/util.go", 
        "Auto-merging pkg/build/apis/build/validation/validation.go", 
        "Auto-merging Godeps/Godeps.json", 
        "Automatic merge failed; fix conflicts and then commit the result."
    ], 
```

meanwhile we'll try one more time:
[merge]

**Comment by bparees** [comment]
ah, now the conflict is showing up.

**Comment by bparees** [comment]
@jim-minter bump for rebase

**Comment by jim-minter** [comment]
@bparees pushed, thanks

**Comment by bparees** [comment]
[merge]

**Pull request merged**